engines/agi/loader_gal_a2.cpp

```diff
@@ -0,0 +1,292 @@
+			continue;
+		}
+
+		// read image as disk one
+		if (readDiskOne(*stream, logicDir)) {
+			debugC(3, "GalLoader_A2: disk one found: %s", imageFile.baseName().c_str());
+			_disks.resize(GAL_A2_DISK_COUNT);
+			_disks[0] = stream;
+			break;
+		} else {
+			delete stream;
+		}
+	}
+
+	// if disk one wasn't found, we're done
+	if (_disks.empty()) {
+		warning("GalLoader_A2: disk one not found");
+		return;
+	}
+
+	// find all other disks by comparing their contents to the logic directory.
+	int disksFound = 1;
+	for (uint i = 1; i < imageFiles.size() && disksFound < GAL_A2_DISK_COUNT; i++) {
+		uint imageFileIndex = (diskOneIndex + i) % imageFiles.size();
+		Common::Path &imageFile = imageFiles[imageFileIndex];
+
+		Common::SeekableReadStream *stream = openA2DiskImage(imageFile, fileMap[imageFile]);
+		if (stream == nullptr) {
+			continue;
+		}
+
+		// check each disk
+		bool diskFound = false;
+		for (int d = 1; d < GAL_A2_DISK_COUNT; d++) {
+			// has disk already been found?
+			if (_disks[d] != nullptr) {
+				continue;
+			}
+
+			if (validateDisk(*stream, d, logicDir)) {
+				_disks[d] = stream;
+				disksFound++;
+				diskFound = true;
+				break;
+			}
+		}
+
+		if (!diskFound) {
+			delete stream;
+		}
+	}
+}
+
+/**
+ * Reads a disk image as disk one by attempting to parse the logic directory
+ * and then validating that all the expected logic resources exist.
+ */
+bool GalLoader_A2::readDiskOne(Common::SeekableReadStream &disk, AgiDir *logicDir) {
+	disk.seek(GAL_A2_LOGDIR_POSITION);
+
+	// attempt to read logic directory
+	for (int i = 0; i < GAL_A2_LOGIC_COUNT; i++) {
+		if (!readDirectoryEntry(disk, logicDir[i])) {
+			return false;
+		}
+	}
+
+	// validate that all disk one logics exist
+	return validateDisk(disk, 0, logicDir);
+}
+
+/**
+ * Reads a directory entry.
+ */
+bool GalLoader_A2::readDirectoryEntry(Common::SeekableReadStream &stream, AgiDir &dirEntry) {
+	// GAL A2 DIR format:
+	// track   8 bits
+	// disk    4 bits (0 for all disks, else 1-3)
+	// sector  4 bits
+	// offset  8 bits
+	byte b0 = stream.readByte();
+	byte b1 = stream.readByte();
+	byte b2 = stream.readByte();
+
+	byte disk = b1 >> 4;
+	byte sector = b1 & 0x0f;
+	uint32 position = A2_DISK_POSITION(b0, sector, b2);
+
+	// use the first disk for resources that are on all disks
+	if (disk > 0) {
+		disk--;
+	}
+
+	// validate entry
+	if (!(disk <= 2 && position < A2_DISK_SIZE)) {
+		return false;
+	}
+
+	dirEntry.volume = disk;
+	dirEntry.offset = position;
+	return true;
+}
+
+/**
+ * Tests if a disk contains all of the expected logic resources.
+ */
+bool GalLoader_A2::validateDisk(Common::SeekableReadStream &disk, byte diskIndex, AgiDir *logicDir) {
+	for (int i = 0; i < GAL_A2_LOGIC_COUNT; i++) {
+		// only validate logics on this disk
+		if (logicDir[i].volume != diskIndex) {
+			continue;
+		}
+
+		// skip logic 64 because it contains incorrect header values
+		if (i == 64) {
+			continue;
+		}
```

Comment:
How does the original handle this case? Is this considered a workaround?

Reply:
This code is Apple II disk image detection, which is something the original didn't do. This does not skip loading logic 64, it skips using it when validating a disk's contents against the expected locations of logic resources according to disk one. It skips it because it has a head value that's off by one, which is not a problem for the original's logic parser (and won't be for ours either once it's written) but is a problem for my scheme of detecting a logic resource by testing that the header lengths match.